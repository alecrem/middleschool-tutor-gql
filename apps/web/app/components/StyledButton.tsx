import type { ButtonHTMLAttributes, ReactNode } from 'react';
import type { LucideIcon } from 'lucide-react';
import { useThemedStyles, useDesignTokens } from '../hooks/useTheme';
import type { ComponentSize, } from '../lib/theme';
import { Icon } from './Icon';

export type ButtonVariant = 'primary' | 'secondary' | 'disabled';
export type ButtonSize = ComponentSize;

interface StyledButtonProps extends Omit<ButtonHTMLAttributes<HTMLButtonElement>, 'disabled'> {
  variant?: ButtonVariant;
  size?: ButtonSize;
  children: ReactNode;
  disabled?: boolean;
  fullWidth?: boolean;
  icon?: LucideIcon;
  iconPosition?: 'left' | 'right';
}

export function StyledButton({
  variant = 'primary',
  size = 'md',
  children,
  disabled = false,
  fullWidth = false,
  icon,
  iconPosition = 'left',
  style: customStyle,
  onMouseEnter,
  onMouseLeave,
  ...props
}: StyledButtonProps) {
  const { colors, utilities } = useThemedStyles();
  const tokens = useDesignTokens();

  const getBaseStyles = () => ({
    border: 'none',
    borderRadius: utilities.borderRadius('md'),
    fontWeight: tokens.typography.fontWeight.medium,
    cursor: disabled ? 'not-allowed' : 'pointer',
    transition: utilities.transition('all'),
    opacity: disabled ? 0.6 : 1,
    width: fullWidth ? '100%' : 'auto',
    outline: 'none',
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'center',
  });

  const getSizeStyles = () => {
    switch (size) {
      case 'xs':
        return {
          padding: `${utilities.spacing('xs')} ${utilities.spacing('sm')}`,
          fontSize: utilities.fontSize('xs'),
          minHeight: '1.5rem',
        };
      case 'sm':
        return {
          padding: `${utilities.spacing('sm')} ${utilities.spacing('md')}`,
          fontSize: utilities.fontSize('sm'),
          minHeight: '2rem',
        };
      case 'lg':
        return {
          padding: `${utilities.spacing('lg')} ${utilities.spacing('xl')}`,
          fontSize: utilities.fontSize('lg'),
          minHeight: '3rem',
        };
      case 'xl':
        return {
          padding: `${utilities.spacing('xl')} ${utilities.spacing('2xl')}`,
          fontSize: utilities.fontSize('xl'),
          minHeight: '3.5rem',
        };
      default: // md
        return {
          padding: `${utilities.spacing('md')} ${utilities.spacing('lg')}`,
          fontSize: utilities.fontSize('base'),
          minHeight: '2.5rem',
        };
    }
  };

  const getVariantStyles = () => {
    if (disabled) {
      return {
        backgroundColor: colors.button.disabled,
        color: colors.button.text,
      };
    }

    switch (variant) {
      case 'secondary':
        return {
          backgroundColor: colors.background.primary,
          color: colors.text.primary,
          border: `1px solid ${colors.border.primary}`,
        };
      default:
        return {
          backgroundColor: colors.button.primary,
          color: colors.button.text,
        };
    }
  };

  const getHoverStyles = () => {
    if (disabled) return {};
    
    switch (variant) {
      case 'secondary':
        return {
          backgroundColor: colors.background.secondary,
        };
      default:
        return {
          opacity: '0.9',
        };
    }
  };

  const combinedStyles = {
    ...getBaseStyles(),
    ...getSizeStyles(),
    ...getVariantStyles(),
    ...customStyle,
  };

  const handleMouseEnter = (e: React.MouseEvent<HTMLButtonElement>) => {
    if (!disabled) {
      const hoverStyles = getHoverStyles();
      Object.assign(e.currentTarget.style, hoverStyles);
    }
    onMouseEnter?.(e);
  };

  const handleMouseLeave = (e: React.MouseEvent<HTMLButtonElement>) => {
    if (!disabled) {
      const originalStyles = getVariantStyles();
      Object.assign(e.currentTarget.style, originalStyles);
    }
    onMouseLeave?.(e);
  };

  const getIconSize = () => {
    switch (size) {
      case 'xs':
      case 'sm':
        return 'xs';
      case 'lg':
      case 'xl':
        return 'sm';
      default:
        return 'xs';
    }
  };

  const renderContent = () => {
    if (!icon) {
      return children;
    }

    const iconElement = <Icon icon={icon} size={getIconSize()} />;
    const gap = utilities.spacing('xs');

    if (iconPosition === 'right') {
      return (
        <span style={{ display: 'flex', alignItems: 'center', gap }}>
          {children}
          {iconElement}
        </span>
      );
    }

    return (
      <span style={{ display: 'flex', alignItems: 'center', gap }}>
        {iconElement}
        {children}
      </span>
    );
  };

  return (
    <button
      style={combinedStyles}
      disabled={disabled}
      onMouseEnter={handleMouseEnter}
      onMouseLeave={handleMouseLeave}
      {...props}
    >
      {renderContent()}
    </button>
  );
}