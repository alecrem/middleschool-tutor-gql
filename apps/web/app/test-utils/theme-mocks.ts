import { vi } from "vitest";

/**
 * Mock implementation for useTheme hooks used in component tests
 * This provides consistent theme values for testing components that depend on theme context
 */
export const mockUseTheme = () => {
  // Mock the ThemeContext first
  vi.mock("../contexts/ThemeContext", () => ({
    useTheme: () => ({
      mode: "light",
      colors: {
        button: {
          primary: "#3B82F6",
          text: "#FFFFFF",
          disabled: "#9CA3AF",
        },
        background: {
          primary: "#FFFFFF",
          secondary: "#F3F4F6",
          card: "#FFFFFF",
        },
        text: {
          primary: "#111827",
          secondary: "#6B7280",
          link: "#3B82F6",
          error: "#EF4444",
        },
        border: {
          primary: "#D1D5DB",
          error: "#EF4444",
        },
      },
      setTheme: vi.fn(),
    }),
  }));

  // Mock the useTheme hooks
  vi.mock("../hooks/useTheme", () => ({
    useThemedStyles: () => ({
      colors: {
        button: {
          primary: "#3B82F6",
          text: "#FFFFFF",
          disabled: "#9CA3AF",
        },
        background: {
          primary: "#FFFFFF",
          secondary: "#F3F4F6",
          card: "#FFFFFF",
        },
        text: {
          primary: "#111827",
          secondary: "#6B7280",
          link: "#3B82F6",
          error: "#EF4444",
        },
        border: {
          primary: "#D1D5DB",
          error: "#EF4444",
        },
      },
      utilities: {
        borderRadius: (_size: string) => "8px",
        transition: (prop: string) => `${prop} 0.2s ease`,
        spacing: (size: string) => {
          const sizes = {
            xs: "4px",
            sm: "8px",
            md: "12px",
            lg: "16px",
            xl: "20px",
            "2xl": "24px",
          };
          return sizes[size as keyof typeof sizes] || "12px";
        },
        fontSize: (size: string) => {
          const sizes = {
            xs: "12px",
            sm: "14px",
            base: "16px",
            lg: "18px",
            xl: "20px",
          };
          return sizes[size as keyof typeof sizes] || "16px";
        },
      },
    }),
    useDesignTokens: () => ({
      typography: {
        fontWeight: {
          medium: "500",
        },
      },
    }),
    useTheme: () => ({
      mode: "light",
      colors: {
        button: {
          primary: "#3B82F6",
          text: "#FFFFFF",
          disabled: "#9CA3AF",
        },
        background: {
          primary: "#FFFFFF",
          secondary: "#F3F4F6",
          card: "#FFFFFF",
        },
        text: {
          primary: "#111827",
          secondary: "#6B7280",
          link: "#3B82F6",
          error: "#EF4444",
        },
        border: {
          primary: "#D1D5DB",
          error: "#EF4444",
        },
      },
      setTheme: vi.fn(),
    }),
  }));
};
