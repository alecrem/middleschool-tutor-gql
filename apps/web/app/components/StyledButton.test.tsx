import { describe, test, expect, vi } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { Search, Download } from "lucide-react";
import { StyledButton } from "./StyledButton";
import { mockUseTheme } from "../test-utils/theme-mocks";

// Mock the theme hooks since they depend on context providers
mockUseTheme();

describe("StyledButton", () => {
  describe("Rendering", () => {
    test("renders button with children", () => {
      render(<StyledButton>Click me</StyledButton>);
      expect(
        screen.getByRole("button", { name: "Click me" })
      ).toBeInTheDocument();
    });

    test("renders with custom className", () => {
      render(<StyledButton className="custom-class">Button</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toHaveClass("custom-class");
    });

    test("renders with custom style", () => {
      render(
        <StyledButton style={{ marginTop: "10px" }}>Styled Button</StyledButton>
      );
      const button = screen.getByRole("button");
      expect(button).toHaveStyle({ marginTop: "10px" });
    });
  });

  describe("Variants", () => {
    test("renders primary variant by default", () => {
      render(<StyledButton>Primary</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toHaveStyle({
        backgroundColor: "#3B82F6",
        color: "#FFFFFF",
      });
    });

    test("renders secondary variant", () => {
      render(<StyledButton variant="secondary">Secondary</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toHaveStyle({
        backgroundColor: "#FFFFFF",
        color: "#111827",
        border: "1px solid #D1D5DB",
      });
    });

    test("renders disabled variant when disabled prop is true", () => {
      render(<StyledButton disabled>Disabled</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toHaveStyle({
        backgroundColor: "#9CA3AF",
        opacity: "0.6",
        cursor: "not-allowed",
      });
    });
  });

  describe("Sizes", () => {
    test("renders medium size by default", () => {
      render(<StyledButton>Medium</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toHaveStyle({
        padding: "12px 16px",
        fontSize: "16px",
        minHeight: "2.5rem",
      });
    });

    test("renders extra small size", () => {
      render(<StyledButton size="xs">Extra Small</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toHaveStyle({
        padding: "4px 8px",
        fontSize: "12px",
        minHeight: "1.5rem",
      });
    });

    test("renders small size", () => {
      render(<StyledButton size="sm">Small</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toHaveStyle({
        padding: "8px 12px",
        fontSize: "14px",
        minHeight: "2rem",
      });
    });

    test("renders large size", () => {
      render(<StyledButton size="lg">Large</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toHaveStyle({
        padding: "16px 20px",
        fontSize: "18px",
        minHeight: "3rem",
      });
    });

    test("renders extra large size", () => {
      render(<StyledButton size="xl">Extra Large</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toHaveStyle({
        padding: "20px 24px",
        fontSize: "20px",
        minHeight: "3.5rem",
      });
    });
  });

  describe("Disabled State", () => {
    test("button is disabled when disabled prop is true", () => {
      render(<StyledButton disabled>Disabled Button</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toBeDisabled();
      expect(button).toHaveStyle({
        opacity: "0.6",
        cursor: "not-allowed",
      });
    });

    test("button is enabled by default", () => {
      render(<StyledButton>Enabled Button</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toBeEnabled();
    });
  });

  describe("Full Width", () => {
    test("renders full width when fullWidth is true", () => {
      render(<StyledButton fullWidth>Full Width</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toHaveStyle({ width: "100%" });
    });

    test("renders auto width by default", () => {
      render(<StyledButton>Auto Width</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toHaveStyle({ width: "auto" });
    });
  });

  describe("Icons", () => {
    test("renders with left icon by default", () => {
      render(<StyledButton icon={Search}>Search</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toBeInTheDocument();
      // Icon should be rendered before text
      const spans = button.querySelectorAll("span");
      expect(spans).toHaveLength(1);
    });

    test("renders with right icon", () => {
      render(
        <StyledButton icon={Download} iconPosition="right">
          Download
        </StyledButton>
      );
      const button = screen.getByRole("button");
      expect(button).toBeInTheDocument();
      // Icon should be rendered after text
      const spans = button.querySelectorAll("span");
      expect(spans).toHaveLength(1);
    });

    test("renders without icon when icon prop is not provided", () => {
      render(<StyledButton>No Icon</StyledButton>);
      const button = screen.getByRole("button");
      expect(button.textContent).toBe("No Icon");
      // No span wrapper should be present
      const spans = button.querySelectorAll("span");
      expect(spans).toHaveLength(0);
    });
  });

  describe("User Interactions", () => {
    test("calls onClick when clicked", async () => {
      const handleClick = vi.fn();
      const user = userEvent.setup();

      render(<StyledButton onClick={handleClick}>Click me</StyledButton>);

      await user.click(screen.getByRole("button"));
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    test("does not call onClick when disabled", async () => {
      const handleClick = vi.fn();
      const user = userEvent.setup();

      render(
        <StyledButton onClick={handleClick} disabled>
          Disabled
        </StyledButton>
      );

      await user.click(screen.getByRole("button"));
      expect(handleClick).not.toHaveBeenCalled();
    });

    test("calls onMouseEnter and onMouseLeave", () => {
      const handleMouseEnter = vi.fn();
      const handleMouseLeave = vi.fn();

      render(
        <StyledButton
          onMouseEnter={handleMouseEnter}
          onMouseLeave={handleMouseLeave}
        >
          Hover me
        </StyledButton>
      );

      const button = screen.getByRole("button");
      fireEvent.mouseEnter(button);
      expect(handleMouseEnter).toHaveBeenCalledTimes(1);

      fireEvent.mouseLeave(button);
      expect(handleMouseLeave).toHaveBeenCalledTimes(1);
    });

    test("disabled button does not receive mouse events", () => {
      const handleMouseEnter = vi.fn();

      render(
        <StyledButton disabled onMouseEnter={handleMouseEnter}>
          Disabled Hover
        </StyledButton>
      );

      const button = screen.getByRole("button");
      fireEvent.mouseEnter(button);
      // Disabled buttons don't receive mouse events in most browsers
      expect(handleMouseEnter).not.toHaveBeenCalled();
    });
  });

  describe("Accessibility", () => {
    test("supports keyboard navigation", async () => {
      const handleClick = vi.fn();
      const user = userEvent.setup();

      render(<StyledButton onClick={handleClick}>Keyboard</StyledButton>);

      const button = screen.getByRole("button");
      button.focus();
      expect(button).toHaveFocus();

      await user.keyboard("{Enter}");
      expect(handleClick).toHaveBeenCalledTimes(1);

      await user.keyboard(" ");
      expect(handleClick).toHaveBeenCalledTimes(2);
    });

    test("has proper ARIA attributes when disabled", () => {
      render(<StyledButton disabled>ARIA Test</StyledButton>);
      const button = screen.getByRole("button");
      expect(button).toHaveAttribute("disabled");
    });
  });

  describe("HTML Attributes", () => {
    test("forwards HTML button attributes", () => {
      render(
        <StyledButton
          type="submit"
          form="test-form"
          data-testid="submit-button"
        >
          Submit
        </StyledButton>
      );

      const button = screen.getByRole("button");
      expect(button).toHaveAttribute("type", "submit");
      expect(button).toHaveAttribute("form", "test-form");
      expect(button).toHaveAttribute("data-testid", "submit-button");
    });
  });
});

describe("StyledButton Snapshots", () => {
  test("matches snapshot for default primary button", () => {
    const { container } = render(<StyledButton>Primary Button</StyledButton>);
    expect(container.firstChild).toMatchSnapshot();
  });

  test("matches snapshot for secondary button with icon", () => {
    const { container } = render(
      <StyledButton variant="secondary" icon={Search} size="lg">
        Search Button
      </StyledButton>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("matches snapshot for disabled button", () => {
    const { container } = render(
      <StyledButton disabled fullWidth>
        Disabled Full Width
      </StyledButton>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test("matches snapshot for extra small button with right icon", () => {
    const { container } = render(
      <StyledButton size="xs" icon={Download} iconPosition="right">
        Download
      </StyledButton>
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
