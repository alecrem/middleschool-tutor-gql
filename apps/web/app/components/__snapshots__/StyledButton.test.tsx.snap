// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StyledButton Snapshots > matches snapshot for default primary button 1`] = `
<button
  style="border-radius: 8px; font-weight: 500; cursor: pointer; transition: all 0.2s ease; opacity: 1; width: auto; outline: none; display: inline-flex; align-items: center; justify-content: center; padding: 12px 16px; font-size: 16px; min-height: 2.5rem; background-color: rgb(59, 130, 246); color: rgb(255, 255, 255);"
>
  Primary Button
</button>
`;

exports[`StyledButton Snapshots > matches snapshot for disabled button 1`] = `
<button
  disabled=""
  style="border-radius: 8px; font-weight: 500; cursor: not-allowed; transition: all 0.2s ease; opacity: 0.6; width: 100%; outline: none; display: inline-flex; align-items: center; justify-content: center; padding: 12px 16px; font-size: 16px; min-height: 2.5rem; background-color: rgb(156, 163, 175); color: rgb(255, 255, 255);"
>
  Disabled Full Width
</button>
`;

exports[`StyledButton Snapshots > matches snapshot for extra small button with right icon 1`] = `
<button
  style="border-radius: 8px; font-weight: 500; cursor: pointer; transition: all 0.2s ease; opacity: 1; width: auto; outline: none; display: inline-flex; align-items: center; justify-content: center; padding: 4px 8px; font-size: 12px; min-height: 1.5rem; background-color: rgb(59, 130, 246); color: rgb(255, 255, 255);"
>
  <span
    style="display: flex; align-items: center; gap: 4px;"
  >
    Download
    <svg
      aria-hidden="true"
      class="lucide lucide-download"
      fill="none"
      height="12"
      stroke="#111827"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      style="width: 12px; height: 12px; color: rgb(17, 24, 39); display: inline-block; vertical-align: middle; flex-shrink: 0; transition: color 0.2s ease;"
      viewBox="0 0 24 24"
      width="12"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M12 15V3"
      />
      <path
        d="M21 15v4a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2v-4"
      />
      <path
        d="m7 10 5 5 5-5"
      />
    </svg>
  </span>
</button>
`;

exports[`StyledButton Snapshots > matches snapshot for secondary button with icon 1`] = `
<button
  style="border: 1px solid rgb(209, 213, 219); border-radius: 8px; font-weight: 500; cursor: pointer; transition: all 0.2s ease; opacity: 1; width: auto; outline: none; display: inline-flex; align-items: center; justify-content: center; padding: 16px 20px; font-size: 18px; min-height: 3rem; background-color: rgb(255, 255, 255); color: rgb(17, 24, 39);"
>
  <span
    style="display: flex; align-items: center; gap: 4px;"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-search"
      fill="none"
      height="16"
      stroke="#111827"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      style="width: 16px; height: 16px; color: rgb(17, 24, 39); display: inline-block; vertical-align: middle; flex-shrink: 0; transition: color 0.2s ease;"
      viewBox="0 0 24 24"
      width="16"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m21 21-4.34-4.34"
      />
      <circle
        cx="11"
        cy="11"
        r="8"
      />
    </svg>
    Search Button
  </span>
</button>
`;
